File: app/android/app/src/main/java/com/gonezo/multiplatform/audioextraction/application/ExecutionPlanner.java
package com.gonezo.multiplatform.audioextraction.application;

import com.gonezo.multiplatform.audioextraction.contract.ExtractionRequest;
import com.gonezo.multiplatform.audioextraction.domain.model.ExecutionPlan;
import com.gonezo.multiplatform.audioextraction.domain.schema.OutputSchema;
import java.util.ArrayList;
import java.util.List;

public class ExecutionPlanner {
  public ExecutionPlan plan(ExtractionRequest request, OutputSchema schema) {
    List<String> required = new ArrayList<>();
    List<String> optional = new ArrayList<>();

    for (String fieldName : schema.fields().keySet()) {
      if (schema.fields().get(fieldName).required()) {
        required.add(fieldName);
      } else {
        optional.add(fieldName);
      }
    }

    boolean includeTranscript = request.options() != null && Boolean.TRUE.equals(request.options().includeTranscript());
    return new ExecutionPlan(required, optional, includeTranscript);
  }
}

File: core/src/main/kotlin/com/gonezo/audioextraction/application/pipeline/ExecutionPlanner.kt
package com.gonezo.audioextraction.application.pipeline

import com.gonezo.audioextraction.domain.contract.ExtractionRequest
import com.gonezo.audioextraction.domain.model.ExecutionPlan
import com.gonezo.audioextraction.domain.schema.OutputSchema

open class ExecutionPlanner {
  open fun plan(request: ExtractionRequest, schema: OutputSchema): ExecutionPlan {
    val required = mutableListOf<String>()
    val optional = mutableListOf<String>()

    for ((fieldName, fieldSchema) in schema.fields) {
      if (fieldSchema.required) {
        required.add(fieldName)
      } else {
        optional.add(fieldName)
      }
    }

    val includeTranscript = request.options?.includeTranscript == true
    return ExecutionPlan(required, optional, includeTranscript)
  }
}
